packages/react-dom/src/client/ReactDOMInput.js

```diff
@@ -212,35 +228,52 @@ export function postMountWrapper(
 ) {
   const node = ((element: any): InputWithWrapperState);
 
+  // Do not assign value if it is already set. This prevents user text input
+  // from being lost during SSR hydration.
   if (props.hasOwnProperty('value') || props.hasOwnProperty('defaultValue')) {
+    const initialValue = node._wrapperState.initialValue;
+    const currentValue = node.value;
+    const value = disableInputAttributeSyncing
+      ? getToStringValue(props.value)
+      : initialValue;
```

Comment:
Seems like previous we didn't have to call this earlier than next condition. What changed? Let's avoid adding more eager computation.

Reply:
Something can be done here. Working on it.

Reply:
I was! I did this in addffb81f